IPython/core/inputtransformer2.py

```diff
@@ -481,6 +481,18 @@ def make_tokens_by_line(lines):
     if not tokens_by_line[-1]:
         tokens_by_line.pop()
 
+    # Convert if using cpython tokenize
+    # upstream bug was fixed in Python 3.7.1, so once we drop 3.7 this can likely be removed. 
+    if (list(map(lambda x: x.type, tokens_by_line[-1])) == 
+            [tokenize.DEDENT] * (len(tokens_by_line[-1]) - 1) + [tokenize.ENDMARKER]):
```

Comment:
Not quite as precise, because it ignores the order, but maybe we could simplify the check using sets:

```python
if {t.type for t in tokens_by_line[-1]} == {tokenize.DEDENT, tokenize.ENDMARKER}:
```

Reply:
If you want to keep the oder I believe you can compare list comprehensions:
```
[t.type for t in tokens_by_line[-1]][-2:] == [tokenize.DEDENT, tokenize.ENDMARKER]
```

```
In [4]: [0, 1,2][-2:] == [1, 2]
Out[4]: True
```

Reply:
That's a looser check, though - you're ignoring anything but the last two things in the line.